arangod/Aql/GraphOptimizerRules.cpp

```diff
@@ -0,0 +1,206 @@
+        if (node->type == NODE_TYPE_REFERENCE ||
+            node->type == NODE_TYPE_VARIABLE) {
+          // we are on the bottom of the tree. Check if it is our pathVar
+          auto variable = static_cast<Variable*>(node->getData());
+          if (pathVariables.contains(variable)) {
+            currentState = PathAccessState::ACCESS_EDGES_OR_VERTICES;
+            matchingPath = variable;
+          }
+        }
+        // Keep for now
+        return node;
+      }
+      case PathAccessState::ACCESS_EDGES_OR_VERTICES: {
+        // we have var.<this-here>
+        // Only vertices || edges supported
+        if (node->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
+          if (node->stringEquals(StaticStrings::GraphQueryEdges)) {
+            isEdgeAccess = true;
+            currentState = PathAccessState::SPECIFIC_DEPTH_ACCESS_DEPTH;
+            return node;
+          } else if (node->stringEquals(StaticStrings::GraphQueryVertices)) {
+            isEdgeAccess = false;
+            currentState = PathAccessState::SPECIFIC_DEPTH_ACCESS_DEPTH;
+            return node;
+          }
+        }
+        // Incorrect type, do not touch, abort this search,
+        // setup next one
+        currentState = PathAccessState::SEARCH_ACCESS_PATH;
+        return node;
+      }
+      case PathAccessState::SPECIFIC_DEPTH_ACCESS_DEPTH: {
+        // we have var.edges[<this-here>], we can only let -1 pass here for
+        // optimization
+        if (node->value.type == VALUE_TYPE_INT &&
+            node->value.value._int == -1) {
+          currentState = PathAccessState::SPECIFIC_DEPTH_ACCESS_INDEXED_ACCESS;
+          return node;
+        }
+        // Incorrect type, do not touch, abort this search,
+        // setup next one
+        currentState = PathAccessState::SEARCH_ACCESS_PATH;
+        return node;
+      }
+      case PathAccessState::SPECIFIC_DEPTH_ACCESS_INDEXED_ACCESS: {
+        // We are in xxxx[-1] pattern, the next node has to be the indexed
+        // access.
+        if (node->type == NODE_TYPE_INDEXED_ACCESS) {
+          // Reset pattern, we can now restart the search
+          currentState = PathAccessState::SEARCH_ACCESS_PATH;
+
+          // Let's switch!!
+          appliedAChange = true;
+          // We searched for the very same variable above!
+          auto const& matchingElements = pathVariables.find(matchingPath);
+          return ast->createNodeReference(isEdgeAccess
+                                              ? matchingElements->second.second
+                                              : matchingElements->second.first);
+        }
+        currentState = PathAccessState::SEARCH_ACCESS_PATH;
+        return node;
+      }
+    }
+  };
+
+  auto newCondition = Ast::traverseAndModify(condition, searchAccessPattern);
+  if (newCondition != condition) {
+    // Swap out everything.
+    return std::make_pair(appliedAChange, newCondition);
+  }
+
+  return std::make_pair(appliedAChange, nullptr);
```

Comment:
As far as I can see, `appliedAChange` should always be false here if we get here. 
Let's add an assertion here, i.e. `TRI_ASSERT(!appliedAChange);`.

Reply:
How can it be the same pointer if it points to a different (i.e. modified) expression?

Reply:
Ok, got it. You are calling the `traverseAndModify(...)` variant that internally uses `changeMember()` to modify the Ast in place.